zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,300 @@
+#![allow(dead_code, unused_variables)]
+use super::Transaction;
+use crate::{
+    block,
+    parameters::{ConsensusBranchId, Network, NetworkUpgrade},
+    serialization::ZcashSerialize,
+    transparent,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use std::io;
+
+const OVERWINTER_VERSION_GROUP_ID: u32 = 0x03C4_8270;
+const SAPLING_VERSION_GROUP_ID: u32 = 0x892F_2085;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+pub const SIGHASH_ALL: u32 = 1;
+const SIGHASH_NONE: u32 = 2;
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub(super) struct SigHasher<'a> {
+    pub(super) trans: &'a Transaction,
+    pub(super) hash_type: u32,
+    pub(super) network: Network,
+    pub(super) height: block::Height,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unreachable!("Zebra checkpoints on Sapling activation"),
+            BeforeOverwinter => unreachable!("Zebra checkpoints on Sapling activation"),
+            Overwinter | Sapling => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Blossom => unimplemented!(),
+            Heartwood => unimplemented!(),
+            Canopy => unimplemented!(),
+        }
+
+        hash.finalize()
+    }
+
+    fn network_upgrade(&self) -> NetworkUpgrade {
+        NetworkUpgrade::current(self.network, self.height)
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade()
+            .branch_id()
+            .expect("Zebra checkpoints on Sapling activation")
+    }
+
+    /// Sighash implementation for the overwinter and sapling consensus branches
+    fn hash_sighash_zip143<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
```

Comment:
ZIP 143 only covers Overwinter, so this function might be better named:
```suggestion
    fn hash_sighash<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
```

ZIP 243 adds the shielded inputs, outputs, and value balance fields for Sapling.

Reply:
I think this comment was made before I filled in the `hash_sighash_zip243` impl but the reason the name includes the zip is because I broke up the entrypoints into the actual algorithms into their own functions rather than putting conditionals within the sighash impl. This was mostly done for readability, so each impl is as simple and consistent with the corresponding zip as possible, rather than branching within the impl and intermixing the logic from both consensus branches.